### ISSUE TITLE
curl 7.75: -c /dev/null deletes character device

### ISSUE THREAD

**Issue by rofl0r** [curl/curl#6882]
### I did this

curl -c /dev/null -C - --retry 5 -k -L http://foo.com/file.tar -o file.tar

### I expected the following

curl would not delete the /dev/null character device, but write/read the cookie jar to/from it

### what happened instead
```
# la /dev/null 
-rw-r--r--    1 root     root            68 Apr 10 10:24 /dev/null
```
i needed to fix the character device afterwards using `rm /dev/null ; mknod /dev/null c 1 3 ; chmod 666 /dev/null`

### curl/libcurl version

[curl -V output]
curl 7.75.0 (x86_64-pc-linux-musl) libcurl/7.75.0 LibreSSL/[redacted]

### operating system
linux

**Comment by bagder** [comment]
Interesting problem. Of course, a normal user is normally not allowed to create or rename files in `/dev` so this is a bit niche.

**Comment by bagder** [comment]
The current method is done to make sure the creation of the file is atomic when multiple processes write to the same file.

What's the correct fix to avoid this issue? Avoid the rename dance if the target file is not a normal file?

**Comment by vixie** [comment]
> ...
> What's the correct fix to avoid this issue? Avoid the rename dance if the target file is not a normal file?

if you want atomic creation then O_CREAT|O_EXCL should give it to you but won't work with character special devices. to get consistent behaviour you'd have to use just O_CREAT|O_NOFOLLOW, then acquire an advisory lock, then fstat() and if S_ISREG() call ftruncate(). there are a _lot_ of security implications around this dance.

**Comment by rofl0r** [comment]
my (probably naive) view is that the procedure should look like this:
read cookies: open(filename, O_RDONLY|O_CLOEXEC), read() ; close();
write cookies: open(filename, O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0660); write() ; close();
in case you want to avoid to lose content written by other curl processes, the write step could maybe use O_APPEND instead of O_WRONLY etc.

**Comment by vixie** [comment]
On Sun, Apr 11, 2021 at 02:22:12AM -0700, rofl0r wrote:
> ...
> read cookies: open(filename, O_RDONLY|O_CLOEXEC), read() ; close();
> write cookies:
> open(filename, O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0660); write(); close();
> ...

i strongly suggest O_NOFOLLOW in both cases.

> in case you want to avoid to lose content written by other curl
> processes, the write step could maybe use O_APPEND instead of O_WRONLY etc.

unless we can define the cooperation model among multiple processes and
ensure that all write()'s are atomic (so, no stdio here), i suggest the
use of advisory locking to either force a delay until the existing process
is complete or induce failure if there's already some process with an open().

-- 
Paul Vixie

**Comment by jay** [comment]
Why isn't it enough to just check if it's a device beforehand? Do we really need to deal with devices atomically?

**Comment by vixie** [comment]
On Sun, Apr 11, 2021 at 09:35:10PM -0700, Jay Satiro wrote:
> Why isn't it enough to just check if it's a device beforehand?
> Do we really need to deal with devices atomically?

if we had a red team they might say there are too many what-ifs
like a path-intermediate symlink changing where it points between
the stat() and later open(). if you open() it first and then
fstat() it there's not a lot that can go wrong. what would be the
advantage of a non-atomic approach?

-- 
Paul Vixie

**Comment by jay** [comment]
In [cookie_output](https://github.com/curl/curl/blob/curl-7_76_0/lib/cookie.c#L1540-L1551) you can see that it uses fopen to create a temporary file and then rename to rename it. What I wanted to know is why we can't just check that if target already exists that it's normal and can be replaced. I wonder why rename was used instead of open O_CREAT O_EXCL like we already do in [tool_create_output_file](https://github.com/curl/curl/blob/curl-7_76_0/src/tool_cb_wrt.c#L78) which is used elsewhere.

**Comment by bagder** [comment]
> why rename was used instead of open O_CREAT O_EXCL

No good reason I think. And as mentioned in here, it should be changed...

**Comment by bagder** [comment]
Well, maybe the fact that `O_EXCL` would change behavior slightly as it would rather make the saving fail rather than overwriting an existing file.

**Comment by jay** [comment]
> No good reason I think. And as mentioned in here, it should be changed...

I edited right as you posted. It appears that there's still the same issue. In other words,

fopen temp file
rename temp file -> target

or

unlink target
open excl target

either way it's the same problem where we don't have an atomic way to figure out if it's a device when we unlink  or rename

**Comment by vixie** [comment]
On Mon, Apr 12, 2021 at 12:28:01AM -0700, Jay Satiro wrote:
> ...or
> 
> unlink target
> open excl target
> 
> either way it's the same problem where we don't have an atomic way
> to figure out if it's a device when we unlink  or rename

this is i think why openat(), unlinkat(), and renameat() were created.
by trimming off the final pathname component and opendir()'ing what's
left, it's possible to avoid the impact of any changes to any interior
symlink in the pathname. under those conditions, if lstat() says that
the final filename (possible target of renameat()) is not a regular
file (so, directory or symlink or device node) then it's safe to fail
and if it is a regular file it's safe to do a renameat().

so, not atomic, but the remaining race condition is that a direntry
that was a regular file one syscall ago has changed to a device node
before the renameat(). i realize that curl (or libcurl) can execute as
root and so stomping /dev/null remains a theoretical possibility. but
unless you're willing to do what some text editors now do, this is "it".

editors, because of symlinks, are sometimes made to copy out the old
file to a backup file, then export the content to the old file, all
without creating or renaming anything. if curl wants to do that with
its cookie jar i don't think there will be complaints.

my recommendation for advisory locking stands, though, in case of multiple
jobs updating the same cookie jar simultaneously. that's not an "atomics"
problem, rather it's a problem caused by some (any?) solutions to the
"atomics" problem.

-- 
Paul Vixie

**Comment by bagder** [comment]
> my recommendation for advisory locking stands, though, in case of multiple jobs updating the same cookie jar simultaneously

This does indeed seem like the most sensible approach that should work with regular files and character devices alike.

The alt-svc and HSTS saving code also have the same flaw as they work the same way as the cookie save.

**Comment by vixie** [comment]
On Mon, Apr 12, 2021 at 04:52:24AM -0700, Daniel Stenberg wrote:
> > my recommendation for advisory locking stands, though, in case of multiple jobs updating the same cookie jar simultaneously
> 
> This does indeed seem like the most sensible approach that should work with regular files and character devices alike.

i worry about the portability layer; vax/vms and ms-dos for example don't have the open() options needed for this.

> The alt-svc and HSTS saving code also have the same flaw as they work the same way as the cookie save.

i can propose a fix if you wish.

**Comment by bagder** [comment]
I started an attempt to use locking instead of rename over in #6884

**Comment by bagder** [comment]
I've not been able to create a fix that doesn't just create another issue (race condition). This issue thus remains.

**Comment by vixie** [comment]
would you mind sharing the roads you tried that went nowhere?

-- 
Paul Vixie

**Comment by bagder** [comment]
The current operation (that this issue refers to) makes one out of multiple processes always successfully write the whole file (and it potentially overwrites the others') since it uses rename at the end of the save.

I have not been able to do open+lock in a way that works that way without a race condition because I need to do open and lock in two separate non-atomic operations!

**Comment by skarnet** [comment]
What semantics do you want for the cookie jar exactly?

If you want to keep the current semantics (one process wins, overwriting all the other changes), it can be achieved with a
sequence of open, blocking lock, write all you need, ftruncate to the current size, close (which unlocks). It does not matter if several processes open the file at the same time: all but one will block on the lock, and the last one to unblock will "win" because it will write() and ftruncate() last.

**Comment by vixie** [comment]
On Mon, May 03, 2021 at 08:35:47AM -0700, Laurent Bercot wrote:
> If you want to keep the current semantics (one process wins, overwriting all the other changes), it can be achieved with a
> sequence of open, blocking lock, write all you need, ftruncate to the current size, close (which unlocks). It does not matter if several processes open the file at the same time: all but one will block on the lock, and the last one to unblock will "win" because it will write() and ftruncate() last.

+1.

-- 
Paul Vixie

**Comment by sr20det** [comment]
or
int fd = open(jar, O_CREAT | O_WRONLY | O_APPEND, 0660);
lockf(fd, F_LOCK, (off_t) strlen(cookie));
write(fd, cookie, strlen(cookie));
close(fd);

**Comment by bagder** [comment]
Oh right the truncate was the missing piece of my previous take. Thanks! Please look at #6884 and see what's right and wrong with that fix!

**Comment by skarnet** [comment]
POSIX says: "Record-locking should not be used in combination with the fopen(), fread(), fwrite(), and other stdio functions. Instead, the more primitive, non-buffered functions (such as open()) should be used."

After opening the file, you're using stdio when writing to it. I don't think it matters because you're locking the whole file every time anyway, but you may want to double-check. And another thing that you should definitely double-check is that your stdio buffers are flushed _before_ the `ftruncate()` 😅

Something that worries me more, however, is a more general problem (I'm unfamiliar with the curl code and only got my first look at it today, so I'm sorry if these concerns have already been raised multiple times and you've already answered them): at build time, you detect functionality availability, but your `#ifdef`s don't have an `#else` clause - so if the functionality is not available on a system, curl will build, but have _incorrect behaviour_. Is that intended?

**Comment by bagder** [comment]
The way I read the code I see no problems with using the f-functions, and I left them as-is to make the change less intrusive. I've made sure to introduce an fflush too.

Yes, this functionality doesn't do any locking if there's no locking function available. Out of most common platforms that'll be Windows, but it also cannot do simultaneous writing to files so should be less of an issue there. There are also platforms without `ftruncate` but I have no idea what to do with this functionality without it...

**Comment by skarnet** [comment]
If only Windows can't lock, then it won't be a problem indeed (but it would be interesting to compare availability of `lockf`-based locks vs. `fcntl`-based ones).

Without `ftruncate`, I suppose you could zero out the rest of the file, or just write a kind of EOF marker, and modify your cookie jar-reading routines accordingly.

**Comment by bagder** [comment]
I share your concern but I have a completely different approach. If I would have to pause and do a **complete** fix for all the lame and incomplete systems out there, then we'd have to spend *significantly* more time and effort for every little change we want to do. And this for fringe and sometimes *theoretical* platforms that may or may even not exist.

Also, writing code and testing for those cases is, while not impossible, really complicated and tricky since we certainly don't run on or have access to any such machines, or even know what they are, so we'd have to create weird simulated solutions to do that. Even more work. And error-prone such.

Different behavior, or perhaps behavior that isn't documented, is a bug. If we find and can reproduce bugs, we work on fixing them. But working our asses off prematurely, is not effective.

Like the `ftruncate` example above. Can you mention a system that lacks the function? (I couldn't find one in my quick gooling the topic) is it worth spending a lot of effort writing up support for that odd case that might not even have a single curl user? Also, even on those systems users could just avoid saving files to the same file name and there won't by any problems. I'm sure users on systems without ftruncate have similar problems all the time...

I'm a pragmatic.

**Comment by jay** [comment]
Just to reiterate I don't see how this is better than checking if the filename is a device beforehand and if so don't unlink. In either case there's potential for a race condition and this way it's at least more straightforward.

**Comment by skarnet** [comment]
What I don't understand, then, is the following: if you don't know of a system without `ftruncate`, then _why do you detect the feature?_ You could just assume it's there (and on POSIX-compliant systems, it will be). And then, if an issue pops up where curl doesn't build on Baddix because Baddix lacks `ftruncate`, _then_ is the time to wonder: is Baddix worth supporting? Can reasonable, non-buggy behaviour be obtained without `ftruncate`, and how much effort does it require? Would people who want to use curl on Baddix be willing to contribute the necessary workaround? etc.

It seems to me that compile-time feature detection is not useful if every platform has the feature and/or lacking it would result in incorrect builds.

That said, I totally share your pain wrt the amount of broken, incomplete systems out there. I have a whole general-purpose library to perform all the system-dependent workarounds I need; I don't even attempt to support Windows, but even getting _the BSDs_ to work is a lot of annoyance.

**Comment by bagder** [comment]
> Just to reiterate I don't see how this is better than checking if the filename is a device beforehand and if so don't unlink. In either case there's potential for a race condition and this way it's at least more straightforward.

One thing the locking improves is for situations where the application only has the permission to write to the file but not to change the directory, so a rename isn't allowed.

**Comment by bagder** [comment]
I've dropped this now and documented it as a known bug.
